File: CSharp/src/Runtime/Scripts/XLoom/loom.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;

namespace FrameworkEase.Universal.X.Utility
{
    public partial class XLoom
#if UNITY_5_3_OR_NEWER
        : UnityEngine.MonoBehaviour
#endif
    {
        #region 业务线程
        internal class Loom : SynchronizationContext
        {
            internal readonly ConcurrentQueue<(SendOrPostCallback, object)> Queue = new();
            public override void Post(SendOrPostCallback callback, object state) { Queue.Enqueue((callback, state)); }
        }
        #endregion

        #region 静态变量
        internal static readonly Mutex initMutex = new();
        internal static int loomCount = 0;
        internal static int loomQueue = 50000;
        internal static Loom[] looms = Array.Empty<Loom>();
        internal static readonly ConcurrentDictionary<Thread, int> loomThreads = new();
        internal static bool[] loomPauses = Array.Empty<bool>();
        #endregion

        #region 初始化
#if UNITY_5_3_OR_NEWER
        internal static XLoom instance;
#endif
        public static void Setup(int count, int step, int queue)
        {
            if (count <= 0 || step <= 0 || queue <= 0) throw new Exception($"XLoom.Setup: invalid parameters, count: {count}, step: {step}, queue: {queue}.");

            lock (initMutex)
            {
#if UNITY_5_3_OR_NEWER
                if (instance == null)
                {
                    var go = new UnityEngine.GameObject("[Loom Updater]");
                    go.AddComponent<XLoom>();
                    DontDestroyOnLoad(go);
                }
#endif

                foreach (var thread in loomThreads.Keys)
                {
                    if (thread == Thread.CurrentThread) continue;
                    try { thread.Interrupt(); }
                    catch { }
                }
                loomThreads.Clear();

                loomCount = count;
                loomQueue = queue;
                looms = new Loom[count];
                loomPauses = new bool[count];

                for (var i = 0; i < count; i++)
                {
                    looms[i] = new Loom();
                    loomPauses[i] = false;
                }

                allTimers = new List<Timer>[count];
                newTimers = new List<Timer>[count];
                removeTimers = new List<int>[count];

                for (var i = 0; i < count; i++)
                {
                    allTimers[i] = new List<Timer>();
                    newTimers[i] = new List<Timer>();
                    removeTimers[i] = new List<int>();
                }

                for (var i = 0; i < count; i++)
                {
                    var loomID = i;
                    var loom = looms[loomID];
#if UNITY_5_3_OR_NEWER
                    if (loomID == 0) loomThreads[Thread.CurrentThread] = loomID;
                    else
#endif
                    {
                        var thread = new Thread(() =>
                        {
                            // 设置业务线程的同步上下文
                            // 通过队列确保异步任务执行的线程一致性
                            SynchronizationContext.SetSynchronizationContext(loom);
                            Loop(loomID, step);
                        });
                        loomThreads[thread] = loomID;
                        thread.Start();
                    }
                }

                XLog.Notice($"XLoom.Setup: allocated {count} loom(s).");
            }
        }

        public static void Reset()
        {
#if UNITY_5_3_OR_NEWER
            if (instance != null)
            {
                UnityEngine.Object.DestroyImmediate(instance.gameObject);
                instance = null;
            }
#else
            lock (initMutex)
            {
                foreach (var thread in loomThreads.Keys)
                {
                    if (thread == Thread.CurrentThread) continue;
                    try { thread.Interrupt(); }
                    catch { }
                }
            }
#endif
            loomThreads.Clear();
            loomCount = 0;
            loomQueue = 0;
            looms = Array.Empty<Loom>();
            loomPauses = Array.Empty<bool>();
            allTimers = null;
            newTimers = null;
            removeTimers = null;
        }
        #endregion

        #region 自循环
#if UNITY_5_3_OR_NEWER
        internal void Awake() { instance = this; }

        internal void Update() { Loop(0, 0); }

        internal void OnDestroy()
        {
            lock (initMutex)
            {
                foreach (var thread in loomThreads.Keys)
                {
                    if (thread == Thread.CurrentThread) continue;
                    try { thread.Interrupt(); }
                    catch { }
                }
            }
            instance = null;
        }
#endif

        internal static void Loop(int loomID, int step)
        {
            try
            {
                while (true)
                {
                    if (!loomPauses[loomID])
                    {
                        while (looms[loomID].Queue.TryDequeue(out var task))
                        {
                            try { task.Item1(task.Item2); }
                            catch (Exception e) { XLog.Error(new Exception($"XLoom.Loop({loomID}): execute runin failed: ", e)); }
                        }
                        try { Tick(loomID); }
                        catch (Exception e) { XLog.Error(new Exception($"XLoom.Loop({loomID}): tick timers failed: ", e)); }
                    }
                    if (step > 0) Thread.Sleep(step);
                    else break;
                }
            }
            catch (Exception e) when (e is OperationCanceledException || e is ThreadInterruptedException || e is ThreadAbortException) { XLog.Notice($"XLoom.Loop({loomID}): receive signal of interrupt."); }
            catch (Exception e) { XLog.Error(new Exception($"XLoom.Loop({loomID}): unexpected exception: ", e)); }
        }
        #endregion

        #region 公共接口
        public static void Pause(int? loomID = null)
        {
            if (loomID != null)
            {
                if (loomID.Value < 0) { XLog.Error($"XLoom.Pause: loom id of {loomID.Value} can not be zero or negative."); return; }
                if (loomID.Value >= loomCount) { XLog.Error($"XLoom.Pause: loom id of {loomID.Value} can not equals or greater than {loomCount}."); return; }
                loomPauses[loomID.Value] = true;
            }
            else for (var id = 0; id < loomPauses.Length; id++) loomPauses[id] = true;
        }

        public static void Resume(int? loomID = null)
        {
            if (loomID != null)
            {
                if (loomID.Value < 0) { XLog.Error($"XLoom.Resume: loom id of {loomID.Value} can not be zero or negative."); return; }
                if (loomID.Value >= loomCount) { XLog.Error($"XLoom.Resume: loom id of {loomID.Value} can not equals or greater than {loomCount}."); return; }
                loomPauses[loomID.Value] = false;
            }
            else for (var id = 0; id < loomPauses.Length; id++) loomPauses[id] = false;
        }

        public static bool RunIn(Action callback, int? loomID = null) { return RunIn(_ => callback(), loomID); }

        public static bool RunIn(SendOrPostCallback callback, int? loomID = null, object state = null)
        {
            if (callback == null) { XLog.Error("XLoom.RunIn: callback can not be nil."); return false; }
            loomID ??= 0;
            if (loomID.Value < 0) { XLog.Error($"XLoom.RunIn: loom id of {loomID.Value} can not be zero or negative."); return false; }
            if (loomID.Value >= loomCount) { XLog.Error($"XLoom.RunIn: loom id of {loomID.Value} can not equals or greater than {loomCount}."); return false; }
            if (looms[loomID.Value].Queue.Count >= loomQueue) { XLog.Error($"XLoom.RunIn: too many runins of {loomID.Value} (queue limit: {loomQueue})."); return false; }
            looms[loomID.Value].Queue.Enqueue((callback, state));
            return true;
        }

        public static int Count => loomCount;

        public static int ID(int? threadID = null)
        {
            if (threadID != null)
            {
                foreach (var thread in loomThreads.Keys)
                {
                    if (thread.ManagedThreadId == threadID) return loomThreads[thread];
                }
                return -1;
            }
            return loomThreads.TryGetValue(Thread.CurrentThread, out var loomID) ? loomID : -1;
        }
        #endregion
    }
}

File: Go/XLoom/loom.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XLoom

import (
	"fmt"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"github.com/frameworkease/Universal.X.Utility/Go/XLog"
	"github.com/illumitacit/gostd/quit"
	"github.com/petermattis/goid"
)

var (
	loomInitMutex sync.Mutex
	loomPause     []bool
	loomPauseSig  []chan bool
	loomInitSig   []chan os.Signal
	loomCloseSig  []chan bool
	loomCloseWait sync.WaitGroup
	loomIDMap     = make(map[int64]int)
	loomIDMu      sync.Mutex
	loomCount     int
	loomTask      []chan func()
)

func Setup(count, step, queue int) {
	loomInitMutex.Lock()
	defer loomInitMutex.Unlock()

	if count <= 0 || step <= 0 || queue <= 0 {
		panic(fmt.Sprintf("XLoom.Setup: invalid parameters, count: %v, step: %v, queue: %v.", count, step, queue))
	}

	// 关闭所有线程。
	if len(loomCloseSig) > 0 {
		for _, ch := range loomCloseSig {
			ch <- true
		}
		loomCloseWait.Wait()
	}
	loomCloseWait = sync.WaitGroup{}

	loomCount = count

	loomTask = make([]chan func(), count)
	loomInitSig = make([]chan os.Signal, count)
	loomCloseSig = make([]chan bool, count)
	loomPause = make([]bool, count)
	loomPauseSig = make([]chan bool, count)

	for i := range count {
		loomTask[i] = make(chan func(), queue)
		loomInitSig[i] = make(chan os.Signal, 1)
		loomCloseSig[i] = make(chan bool, 1)
		loomPauseSig[i] = make(chan bool, 1)
	}

	setupTimer(count)

	wg := sync.WaitGroup{}
	for i := range count {
		wg.Add(1)

		doneOnce := sync.Once{}
		RunAsyncT1(func(loomID int) {
			initSig := loomInitSig[i]
			signal.Notify(initSig, syscall.SIGTERM, syscall.SIGINT)
			pauseSig := loomPauseSig[i]
			closeSig := loomCloseSig[i]

			loomCloseWait.Add(1)
			quit.GetWaiter().Add(1)
			defer func() {
				quit.GetWaiter().Done()
				loomCloseWait.Done()
			}()

			loomIDMu.Lock()
			loomIDMap[goid.Get()] = loomID
			loomIDMu.Unlock()

			updateTicker := time.NewTicker(time.Millisecond * time.Duration(step))
			defer updateTicker.Stop()

			doneOnce.Do(func() { // 确保只调用一次，否则recover后会重复调用
				wg.Done() // 确保线程启动完成
			})

			for {
				if loomPause[loomID] {
					select {
					case <-updateTicker.C:
					case val := <-pauseSig:
						XLog.Notice("XLoom.Loop(%v): receive signal of pause(%v).", loomID, val)
					case <-closeSig:
						XLog.Notice("XLoom.Loop(%v): receive signal of close.", loomID)
						return
					case sig, ok := <-initSig:
						if ok {
							XLog.Notice("XLoom.Loop(%v): receive signal of %v.", i, sig.String())
						} else {
							XLog.Notice("XLoom.Loop(%v): channel of signal is closed.", i)
						}
						return
					case <-quit.GetQuitChannel():
						XLog.Notice("XLoom.Loop(%v): receive signal of quit.", loomID)
						return
					}
				} else {
					select {
					case runIn, ok := <-loomTask[loomID]:
						if ok {
							runIn()
						} else {
							XLog.Error("XLoom.Loop(%v): get runin with ret false.", loomID)
						}
					case <-updateTicker.C:
						updateTimer(loomID)
					case val := <-pauseSig:
						XLog.Notice("XLoom.Loop(%v): receive signal of pause(%v).", loomID, val)
					case <-closeSig:
						XLog.Notice("XLoom.Loop(%v): receive signal of close.", loomID)
						return
					case sig, ok := <-initSig:
						if ok {
							XLog.Notice("XLoom.Loop(%v): receive signal of %v.", i, sig.String())
						} else {
							XLog.Notice("XLoom.Loop(%v): channel of signal is closed.", i)
						}
						return
					case <-quit.GetQuitChannel():
						XLog.Notice("XLoom.Loop(%v): receive signal of quit.", loomID)
						return
					}
				}
			}
		}, i, true)
	}

	XLog.Notice("XLoom.Setup: allocated %v loom(s).", count)
	loomCount = count
	wg.Wait()
}

func Reset() {
	loomInitMutex.Lock()
	defer loomInitMutex.Unlock()

	if len(loomCloseSig) > 0 {
		for _, ch := range loomCloseSig {
			ch <- true
		}
		loomCloseWait.Wait()
	}
	loomCloseWait = sync.WaitGroup{}

	loomCount = 0
	loomTask = nil
	loomInitSig = nil
	loomCloseSig = nil
	loomPause = nil
	loomPauseSig = nil
	loomIDMap = make(map[int64]int)
	loomIDMu = sync.Mutex{}
}

func Pause(loomID ...int) {
	if len(loomID) == 1 {
		lid := loomID[0]
		if lid < 0 {
			XLog.Error("XLoom.Pause: loom id of %v can not be zero or negative.", lid)
			return
		}
		if lid >= loomCount {
			XLog.Error("XLoom.Pause: loom id of %v can not equals or greater than %v.", lid, Count())
			return
		}
		loomPause[lid] = true
		loomPauseSig[lid] <- true
	} else {
		for lid := range loomPause {
			loomPause[lid] = true
			loomPauseSig[lid] <- true
		}
	}
}

func Resume(loomID ...int) {
	if len(loomID) == 1 {
		lid := loomID[0]
		if lid < 0 {
			XLog.Error("XLoom.Resume: loom id of %v can not be zero or negative.", lid)
			return
		}
		if lid >= loomCount {
			XLog.Error("XLoom.Resume: loom id of %v can not equals or greater than %v.", lid, Count())
			return
		}
		loomPause[lid] = false
		loomPauseSig[lid] <- false
	} else {
		for lid := range loomPause {
			loomPause[lid] = false
			loomPauseSig[lid] <- false
		}
	}
}

func RunIn(callback func(), loomID ...int) bool {
	if callback == nil {
		XLog.Error("XLoom.RunIn: callback can not be nil.")
		return false
	}
	lid := -1
	if len(loomID) == 1 {
		lid = loomID[0]
	} else {
		lid = 0
	}
	if lid < 0 {
		XLog.Error("XLoom.RunIn: loom id of %v can not be zero or negative.", lid)
		return false
	}
	if lid >= loomCount {
		XLog.Error("XLoom.RunIn: loom id of %v can not equals or greater than %v.", lid, Count())
		return false
	}
	ch := loomTask[lid]
	select {
	case ch <- callback:
		return true
	default:
		XLog.Error("XLoom.RunIn: too many runins of %v.", lid)
		return false
	}
}

func Count() int { return loomCount }

func ID(goroutineID ...int64) int {
	var gid int64
	if len(goroutineID) == 1 {
		gid = goroutineID[0]
	} else {
		gid = goid.Get()
	}
	if loomID, ok := loomIDMap[gid]; ok {
		return loomID
	}
	return -1
}
